[ews] merge-queue fails to find reviewer on GitHub PRs with lots of review activity

JonWBedard opened the pull request:
#### 42fd020fc0c32ab8ef648cacc49fcbe3d2fd1a81
<pre>
[ews] merge-queue fails to find reviewer on GitHub PRs with lots of review activity
<a href="https://bugs.webkit.org/show_bug.cgi?id=243548">https://bugs.webkit.org/show_bug.cgi?id=243548</a>
rdar://98129710

Reviewed by Aakash Jain.

* Tools/CISupport/ews-build/steps.py:
(GitHubMixin):
(GitHubMixin.get_reviewers): Handle pagination when retrieving reviews. Approval
and rejection should override older approval or rejection.
* Tools/CISupport/ews-build/steps_unittest.py:
(TestGitHubMixin):
(TestGitHubMixin.Response): Mock of request&apos;s response class.
(TestGitHubMixin.test_no_reviewers):
(TestGitHubMixin.test_single_review):
(TestGitHubMixin.test_multipe_reviews):
(TestGitHubMixin.test_retracted_review):
(TestGitHubMixin.test_pagination):
(TestGitHubMixin.test_reviewers_invalid_response):
(TestGitHubMixin.test_reviewers_error):

Canonical link: <a href="https://commits.webkit.org/256752@main">https://commits.webkit.org/256752@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/84ed6b26d08a749ff708e7d1c6de478c4b3e0ab1

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/96746 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/6011 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/29841 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/106274 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/166573 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/100729 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/6226 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/34747 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/89134 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/102983 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/102421 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/4654 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/83362 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/31618 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/86494 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/29841 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/74549 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/49 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/29841 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/20/builds/95701 "Passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/45 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/29841 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/4830 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/83362 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/2250 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/79/builds/1271 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/29841 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
<!--EWS-Status-Bubble-End-->

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -213,17 +214,32 @@ def get_reviewers(self, pr_number, repository_url=None):
+        reviews = []
+        reviews_url = f'{api_url}/pulls/{pr_number}/reviews?per_page={self.LIMIT}'
+        page = 1
+        while page < 10:
```
Let's make this 10 as a parameter as well, so that it's explicit and easy to modify later on if we need to change it.

aj062 replied:
Can change it to something like: `for page in range(1, LIMIT_NUM_PAGES + 1)`

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -162,6 +162,7 @@ class GitHubMixin(object):
     MERGE_QUEUE_LABEL = 'merge-queue'
     UNSAFE_MERGE_QUEUE_LABEL = 'unsafe-merge-queue'
     REQUEST_MERGE_QUEUE_LABEL = 'request-merge-queue'
+    LIMIT = 100
```
Nit: can rename to something like: `LIMIT_REVIEWS_PER_PAGE`

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -213,17 +214,32 @@ def get_reviewers(self, pr_number, repository_url=None):
+            if not isinstance(response_content, list):
+                break
+            reviews += response_content
+            if len(response_content) >= self.LIMIT:
```
why do we need this check?

JonWBedard replied:
We're asking for 100 on the page. If the page has less than 100, we can safely assume this is the last page. Most of the time, we won't even make a second request.

aj062 replied:
Seems like the comparison need to be inverted than. Should be `if len(response_content) < self.LIMIT:`

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -213,17 +214,32 @@ def get_reviewers(self, pr_number, repository_url=None):
+            if not content:
+                break
+            response_content = content.json() or []
+            if not isinstance(response_content, list):
```
would be good to add `self._addToLog` here to help in debugging such failures.

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -213,17 +214,32 @@ def get_reviewers(self, pr_number, repository_url=None):
+            if not reviewer:
+                continue
+            if review.get('state') == 'APPROVED':
+                approved[reviewer][review.get('id', 0)]
```
what are we doing here?

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -213,17 +214,32 @@ def get_reviewers(self, pr_number, repository_url=None):
+            if review.get('state') == 'APPROVED':
+                approved[reviewer][review.get('id', 0)]
+            elif review.get('state') == 'CHANGES_REQUESTED':
+                rejected[reviewer][review.get('id', 0)]
```
Ditto

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -213,17 +214,32 @@ def get_reviewers(self, pr_number, repository_url=None):
+                approved[reviewer][review.get('id', 0)]
+            elif review.get('state') == 'CHANGES_REQUESTED':
+                rejected[reviewer][review.get('id', 0)]
+        return [reviewer for reviewer, id_val in approved.items() if id_val > rejected.get(reviewer, 0)]
```
what's the logic being implemented here? what's `id_val`? maybe rename id_val to something more readable.

JonWBedard replied:
We're trying to make sure that someone can override an r+ with an r- in the future. `id_val` comes from GitHub, it's the review "id" (according to GitHub's API). Larger numbers are newer. We could use timestamp, expect I didn't really want to decode the GitHub timestamp.

aj062 replied:
ok.
Also, id_val is a local variable in this line, maybe we can call it `_id`.
Also, since this is important logic, probably having a explicit for loop with a comment explaining what we are doing is better (instead of using list comprehension).

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -213,17 +215,35 @@ def get_reviewers(self, pr_number, repository_url=None):
+                self._addToLog('stdio', f"Malformed response when listing reviews with '{url}'\n")
+                break
+            reviews += response_content
+            if self.PER_PAGE_LIMIT < len(response_content):
```
This still seems inverse. Maybe I am missing something.

JonWBedard replied:
Nope, my brain just wasn't working...updating

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -213,17 +215,35 @@ def get_reviewers(self, pr_number, repository_url=None):
+                break
+            reviews += response_content
+            if len(response_content) < self.PER_PAGE_LIMIT:
+                continue
```
should it be `continue` or `break`. Seems like continue will result in an infinite loop.

aj062 reviewed (commented):
would be good to add some unit-tests.
Can override/mock GitHubMixin.fetch_data_from_url_with_authentication_github() in unit-tests to avoid network request.

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -213,17 +215,34 @@ def get_reviewers(self, pr_number, repository_url=None):
-            return []
+        reviews = []
+        reviews_url = f'{api_url}/pulls/{pr_number}/reviews?per_page={self.PER_PAGE_LIMIT}'
+        for page in range(1, self.NUM_PAGE_LIMIT):
```
Nit: this range would be one less than self.NUM_PAGE_LIMIT

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -213,17 +215,34 @@ def get_reviewers(self, pr_number, repository_url=None):
-        if not content:
-            return []
+        reviews = []
+        reviews_url = f'{api_url}/pulls/{pr_number}/reviews?per_page={self.PER_PAGE_LIMIT}'
```
do we need to construct this variable here, it's used only to construct the `url` variable. We can directly construct url in one go inside the for loop.

e.g.: `url = f'{api_url}/pulls/{pr_number}/reviews?per_page={self.PER_PAGE_LIMIT}&page={page}`

JonWBedard committed:
[ews] merge-queue fails to find reviewer on GitHub PRs with lots of review activity
https://bugs.webkit.org/show_bug.cgi?id=243548
rdar://98129710

Reviewed by Aakash Jain.

* Tools/CISupport/ews-build/steps.py:
(GitHubMixin):
(GitHubMixin.get_reviewers): Handle pagination when retrieving reviews. Approval
and rejection should override older approval or rejection.
* Tools/CISupport/ews-build/steps_unittest.py:
(TestGitHubMixin):
(TestGitHubMixin.Response): Mock of request's response class.
(TestGitHubMixin.test_no_reviewers):
(TestGitHubMixin.test_single_review):
(TestGitHubMixin.test_multipe_reviews):
(TestGitHubMixin.test_retracted_review):
(TestGitHubMixin.test_pagination):
(TestGitHubMixin.test_reviewers_invalid_response):
(TestGitHubMixin.test_reviewers_error):

Canonical link: https://commits.webkit.org/256752@main
